File: app/src/main/java/dev/bee/kanjianki/backup/DatabaseBackupWorker.java
package dev.bee.kanjianki.backup;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import dev.bee.kanjianki.core.DatabaseBackupPolicy;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;

public final class DatabaseBackupWorker extends Worker {
    private static final String TAG = "DatabaseBackupWorker";

    public DatabaseBackupWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        return doWork(new AndroidBackupEnvironment(getApplicationContext()), System.currentTimeMillis());
    }

    static Result doWork(BackupEnvironment environment, long nowMillis) {
        return backupDatabase(
                environment.databasePath(DatabaseBackupPolicy.DB_NAME),
                environment.filesDir(),
                nowMillis,
                DatabaseBackupWorker::checkpoint,
                DatabaseBackupWorker::copyFile);
    }

    static Result backupDatabase(
            File dbFile,
            File filesDir,
            long nowMillis,
            Checkpointer checkpointer,
            FileCopier copier) {
        if (!dbFile.exists()) {
            return Result.failure();
        }

        File backupDir = DatabaseBackupPolicy.backupDir(filesDir);
        if (!backupDir.exists() && !backupDir.mkdirs()) {
            return Result.failure();
        }

        try {
            checkpointer.checkpoint(dbFile);
        } catch (RuntimeException error) {
            warn("Backup checkpoint failed; copying database without a fresh WAL checkpoint.", error);
        }

        File dest = DatabaseBackupPolicy.backupFile(filesDir, nowMillis);

        try {
            copier.copy(dbFile, dest);
        } catch (IOException e) {
            if (!dest.delete()) {
                warn("Failed to delete incomplete backup: " + dest.getName());
            }
            warn("Database backup failed.", e);
            return Result.failure();
        }

        pruneOldBackups(backupDir);
        return Result.success();
    }

    interface Checkpointer {
        void checkpoint(File dbFile);
    }

    interface FileCopier {
        void copy(File src, File dst) throws IOException;
    }

    interface BackupEnvironment {
        File databasePath(String name);

        File filesDir();
    }

    static BackupEnvironment androidEnvironment(Context context) {
        return new AndroidBackupEnvironment(context);
    }

    private static final class AndroidBackupEnvironment implements BackupEnvironment {
        private final Context context;

        AndroidBackupEnvironment(Context context) {
            this.context = context;
        }

        @Override
        public File databasePath(String name) {
            return context.getDatabasePath(name);
        }

        @Override
        public File filesDir() {
            return context.getFilesDir();
        }
    }

    static void checkpoint(File dbFile) {
        checkpoint(dbFile, DatabaseBackupWorker::openCheckpointDatabase);
    }

    static void checkpoint(File dbFile, CheckpointDatabaseOpener opener) {
        CheckpointDatabase db = null;
        try {
            db = opener.open(dbFile);
            db.checkpoint();
        } catch (IOException | RuntimeException error) {
            warn("Backup checkpoint failed; copying database without a fresh WAL checkpoint.", error);
        } finally {
            closeCheckpointDatabase(db);
        }
    }

    private static CheckpointDatabase openCheckpointDatabase(File dbFile) {
        SQLiteDatabase db = SQLiteDatabase.openDatabase(
                dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
        return new SQLiteCheckpointDatabase(db);
    }

    private static void closeCheckpointDatabase(CheckpointDatabase db) {
        if (db == null) {
            return;
        }
        try {
            db.close();
        } catch (IOException | RuntimeException e) {
            warn("Failed to close database after backup checkpoint.", e);
        }
    }

    interface CheckpointDatabaseOpener {
        CheckpointDatabase open(File dbFile) throws IOException;
    }

    interface CheckpointDatabase {
        void checkpoint() throws IOException;

        void close() throws IOException;
    }

    private static final class SQLiteCheckpointDatabase implements CheckpointDatabase {
        private final SQLiteDatabase db;

        SQLiteCheckpointDatabase(SQLiteDatabase db) {
            this.db = db;
        }

        @Override
        public void checkpoint() {
            db.rawQuery("PRAGMA wal_checkpoint(TRUNCATE)", null).close();
        }

        @Override
        public void close() {
            db.close();
        }
    }

    static void copyFile(File src, File dst) throws IOException {
        try (FileInputStream inStream = new FileInputStream(src);
             FileOutputStream outStream = new FileOutputStream(dst);
             FileChannel inChannel = inStream.getChannel();
             FileChannel outChannel = outStream.getChannel()) {
            long size = inChannel.size();
            long transferred = 0;
            while (transferred < size) {
                transferred += inChannel.transferTo(transferred, size - transferred, outChannel);
            }
            outChannel.force(true);
        }
    }

    static void pruneOldBackups(File backupDir) {
        for (File oldBackup : DatabaseBackupPolicy.oldBackupsToPrune(backupDir)) {
            if (!oldBackup.delete()) {
                warn("Failed to prune old backup: " + oldBackup.getName());
            }
        }
    }

    private static void warn(String message) {
        try {
            Log.w(TAG, message);
        } catch (RuntimeException ignored) {
            // Android Log is unavailable in local JVM tests.
        }
    }

    private static void warn(String message, Throwable error) {
        try {
            Log.w(TAG, message, error);
        } catch (RuntimeException ignored) {
            // Android Log is unavailable in local JVM tests.
        }
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/backup/DatabaseBackupWorker.kt
package dev.bee.kanjianki.backup

import dev.bee.kanjianki.AppLocalStoreFactory

import android.content.Context
import android.os.Build
import android.util.Log
import androidx.work.ListenableWorker.Result
import androidx.work.Worker
import androidx.work.WorkerParameters
import dev.bee.kanjianki.core.DatabaseBackupPolicy
import dev.bee.kanjianki.core.DatabaseBackupAvailabilityPolicy
import dev.bee.kanjianki.data.LocalStore
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.zip.GZIPOutputStream

class DatabaseBackupWorker(
    context: Context,
    workerParams: WorkerParameters,
) : Worker(context, workerParams) {
    override fun doWork(): Result {
        return doWork(
            AndroidBackupEnvironment(applicationContext),
            System.currentTimeMillis(),
            Build.VERSION.SDK_INT,
        )
    }

    /**
     * Produces a WAL-safe backup at [dest] from [dbFile]. Implementations copy committed
     * WAL content (e.g. `VACUUM INTO`) rather than doing a plain file copy that would
     * ignore the `-wal`/`-shm` sidecars and tear or stale the backup.
     */
    fun interface Snapshotter {
        @Throws(IOException::class)
        fun snapshot(dbFile: File, dest: File)
    }

    internal fun interface BackupPublisher {
        @Throws(IOException::class)
        fun publish(partial: File, destination: File)
    }

    interface BackupEnvironment {
        fun databasePath(name: String): File

        fun filesDir(): File

        @Throws(IOException::class)
        fun snapshot(dbFile: File, dest: File)
    }

    private class AndroidBackupEnvironment(
        private val context: Context,
    ) : BackupEnvironment {
        override fun databasePath(name: String): File {
            return context.getDatabasePath(name)
        }

        override fun filesDir(): File {
            return context.filesDir
        }

        override fun snapshot(dbFile: File, dest: File) {
            AppLocalStoreFactory.create(context).use { store ->
                store.snapshotInto(dest)
            }
        }
    }

    companion object {
        private const val TAG = "DatabaseBackupWorker"
        private const val BACKUP_TIMESTAMP_PATTERN = "yyyyMMdd_HHmmss"
        private val BACKUP_SCRATCH_NAME =
            Regex("^kanji_anki_simple_(\\d{8}_\\d{6})\\.db\\.gz\\.(?:tmp|partial|pre-restore\\.tmp)$")

        @JvmStatic
        fun doWork(environment: BackupEnvironment, nowMillis: Long): Result {
            return doWork(
                environment,
                nowMillis,
                DatabaseBackupAvailabilityPolicy.MIN_SAFE_ANDROID_API,
            )
        }

        internal fun doWork(
            environment: BackupEnvironment,
            nowMillis: Long,
            apiLevel: Int,
        ): Result {
            if (!DatabaseBackupAvailabilityPolicy.forAndroidApi(apiLevel).operationsAllowed) {
                // This is a permanent platform capability decision. A stale periodic
                // request should finish successfully without touching existing archives.
                return Result.success()
            }
            return backupDatabase(
                environment.databasePath(DatabaseBackupPolicy.DB_NAME),
                environment.filesDir(),
                nowMillis,
                environment::snapshot,
            )
        }

        @JvmStatic
        fun backupDatabase(
            dbFile: File,
            filesDir: File,
            nowMillis: Long,
            snapshotter: Snapshotter,
        ): Result = backupDatabase(dbFile, filesDir, nowMillis, snapshotter, ::publishAtomically)

        internal fun backupDatabase(
            dbFile: File,
            filesDir: File,
            nowMillis: Long,
            snapshotter: Snapshotter,
            publisher: BackupPublisher,
        ): Result {
            if (!dbFile.exists()) {
                return Result.failure()
            }

            val backupDir = DatabaseBackupPolicy.backupDir(filesDir)
            if ((!backupDir.exists() && !backupDir.mkdirs()) || !backupDir.isDirectory) {
                return Result.failure()
            }
            if (!deleteStaleScratchFiles(backupDir)) {
                return Result.failure()
            }

            val dest = DatabaseBackupPolicy.backupFile(filesDir, nowMillis)
            // Snapshot to a raw uncompressed temp copy, then gzip it into place. SQLite
            // Databases compress ~4-10x, and retention keeps a small tiered set of
            // seven recent daily plus up to four older weekly archives.
            val temp = File(backupDir, dest.name + ".tmp")
            val partial = File(backupDir, dest.name + ".partial")
            if (!deleteIncomplete(temp) || !deleteIncomplete(partial)) {
                return Result.failure()
            }

            try {
                snapshotter.snapshot(dbFile, temp)
                requireSnapshot(temp)
                gzipFile(temp, partial)
                publisher.publish(partial, dest)
            } catch (error: IOException) {
                deleteIncomplete(partial)
                warn(DatabaseBackupPolicy.sanitizedDiagnosticLine("Database backup failed.", error))
                return Result.failure()
            } catch (error: RuntimeException) {
                deleteIncomplete(partial)
                warn(DatabaseBackupPolicy.sanitizedDiagnosticLine("Database backup failed.", error))
                return Result.failure()
            } finally {
                deleteIncomplete(temp)
            }

            pruneOldBackups(backupDir)
            return Result.success()
        }

        @JvmStatic
        @Throws(IOException::class)
        fun gzipFile(src: File, dest: File) {
            FileInputStream(src).use { input ->
                FileOutputStream(dest).use { fileOut ->
                    GZIPOutputStream(fileOut).use { gzip ->
                        input.copyTo(gzip)
                        gzip.finish()
                        // Flush compressed bytes to disk before GZIPOutputStream.close()
                        // closes the underlying stream.
                        fileOut.fd.sync()
                    }
                }
            }
        }

        @Throws(IOException::class)
        internal fun publishAtomically(partial: File, destination: File) {
            check(partial.parentFile == destination.parentFile) {
                "Backup partial and destination must share a directory"
            }
            if (destination.exists()) {
                throw IOException("Backup destination already exists")
            }
            Files.move(
                partial.toPath(),
                destination.toPath(),
                StandardCopyOption.ATOMIC_MOVE,
            )
            val destinationDirectory = destination.parentFile
                ?: throw IOException("Backup destination has no parent directory")
            SystemDirectorySynchronizer.sync(destinationDirectory)
        }

        @JvmStatic
        fun androidEnvironment(context: Context): BackupEnvironment {
            return AndroidBackupEnvironment(context)
        }

        private fun deleteIncomplete(dest: File): Boolean {
            if (dest.exists() && !dest.delete()) {
                warn("Failed to delete incomplete backup: ${dest.name}")
                return false
            }
            return true
        }

        internal fun deleteStaleScratchFiles(backupDir: File): Boolean {
            val entries = backupDir.listFiles()
            if (entries == null) {
                warn("Failed to inspect incomplete backups")
                return false
            }
            for (entry in entries) {
                if (isRecognizedBackupScratch(entry.name) && !deleteIncomplete(entry)) {
                    return false
                }
            }
            return true
        }

        private fun isRecognizedBackupScratch(name: String): Boolean {
            val match = BACKUP_SCRATCH_NAME.matchEntire(name) ?: return false
            val parser = SimpleDateFormat(BACKUP_TIMESTAMP_PATTERN, Locale.US).apply {
                isLenient = false
            }
            return try {
                parser.parse(match.groupValues[1])
                true
            } catch (_: ParseException) {
                false
            }
        }

        @Throws(IOException::class)
        private fun requireSnapshot(snapshot: File) {
            if (!snapshot.isFile || snapshot.length() <= 0L) {
                throw IOException("Snapshot operation produced no database")
            }
        }

        @JvmStatic
        fun pruneOldBackups(backupDir: File) {
            for (oldBackup in DatabaseBackupPolicy.oldBackupsToPrune(backupDir)) {
                if (!oldBackup.delete()) {
                    warn("Failed to prune old backup: ${oldBackup.name}")
                }
            }
        }

        private fun warn(message: String) {
            try {
                Log.w(TAG, message)
            } catch (_: RuntimeException) {
                // Android Log is unavailable in local JVM tests.
            }
        }
    }
}
